src/model.jl

```diff
@@ -97,12 +100,38 @@ function fit_cnmf(data; L=10, K=5, alg=:mult,
         # Record time and loss
         push!(time_hist, time_hist[end] + dur)
         push!(loss_hist, loss)
+
+        # Check convergence
+        if check_convergence && converged(loss_hist, patience, tol)
+            break
+        end
     end
 
     return CNMF_results(data, W, H, time_hist, loss_hist,
                         l1_H, l2_H, l1_W, l2_W, alg)
 end
 
+"""
+Check for model convergence
+"""
+function converged(loss_hist, patience, tol)
+
+    # If we have not run for `patience` iterations,
+    # we have not converged.
+    if length(loss_hist) <= patience
```

Comment:
Is this the best application of `patience`? My thought would be to use it for something like the number of iterations `d_loss` has to be small before exiting (`d_loss` would have to be small for 3 consecutive iterations for example).

Reply:
Yeah that's how I think it currently works. Looking at the code below:

```julia 
d_loss = diff(loss_hist[end-patience:end])

     # Objective converged
    if (all(abs.(d_loss) .< tol))
        return true
    else
        return false
    end
```
we take the first difference across `patience` number of prior loss values and check that they're all small. The check above is because when the array has fewer than `patience` number of entries  the below code would throw an indexing error.